### DISCUSSION TITLE
Seeking help for Compiling FDS 6.9.1 on HPC

### DISCUSSION THREAD

**Discussion by mingcian3** [firemodels/fds#13738] [Q&A] [answered]
Dear All,

I am trying to compile the latest FDS (6.9.1) available on GitHub on the HPC in my university. However, I encountered some problems and the compilation could not proceed. I have some experience to compile FDS on HPC and tried several commands but it does not work for my current compilation. Could you give me some suggestions that I can further try? I list the major steps I’ve tried here. Thank you very much in advance for your suggestions.
--------------------
Step 1:
I fork the ‘exp’, ‘fds’, ‘out’, ‘smv’ in my folder named ‘~/FDS6-9-1/firemodles’

Step 2:
I use ‘dos2unix make_fds.sh’ to convert file make_fds.sh to Unix format.

Step 3-1: impi version (impi/2021.10.0-intel-compilers-2023.2.1)
I load the latest impi module (impi/2021.10.0-intel-compilers-2023.2.1) available on the university HPC. 
I go to ‘~/Build/impi_intel_linux’ and use the command ‘./make_fds.sh’ to compile the FDS. 
However, the error message shows:

../Scripts/build_thirdparty_libs.sh: line 2: $'\r': command not found
../Scripts/build_thirdparty_libs.sh: line 6: syntax error near unexpected token `fi'
'./Scripts/build_thirdparty_libs.sh: line 6: `fi
Building impi_intel_linux
fatal: No names found, cannot describe anything.
/bin/bash: ../Scripts/osx_clt.sh: /bin/bash^M: bad interpreter: No such file or directory
mpiifx -c -m64 -fc=ifort -O2 -ipo -no-wrap-margin -fpp -DGITHASH_PP=\"-master\" -DGITDATE_PP=\""Sat Nov 9 09:28:05 2024 -0500\"" -DBUILDDATE_PP=\""Nov 11, 2024  16:26:01\""    -DUSE_IFPORT  -diag-disable=10448 ../../Source/prec.f90
mpiifx -c -m64 -fc=ifort -O2 -ipo -no-wrap-margin -fpp -DGITHASH_PP=\"-master\" -DGITDATE_PP=\""Sat Nov 9 09:28:05 2024 -0500\"" -DBUILDDATE_PP=\""Nov 11, 2024  16:26:01\""    -DUSE_IFPORT  -diag-disable=10448  ../../Source/imkl.f90
Error: can not overwrite predefined compiler - ifx!
You should use mpiifort
ex. mpiifort -fc=ifort
make: *** [../makefile:151: prec.o] Error 1
make: *** Waiting for unfinished jobs....
Error: can not overwrite predefined compiler - ifx!
You should use mpiifort
ex. mpiifort -fc=ifort
make: *** [../makefile:151: imkl.o] Error 1

Step 3-2: another impi version ( impi/2021.4.0-intel-compilers-2021.4.0)
I tried an older impi module (impi/2021.4.0-intel-compilers-2021.4.0) that I used for compiling FDS 6.8.0. Another error message shows:
../Scripts/build_thirdparty_libs.sh: line 2: $'\r': command not found
../Scripts/build_thirdparty_libs.sh: line 6: syntax error near unexpected token `fi'
'./Scripts/build_thirdparty_libs.sh: line 6: `fi
Building impi_intel_linux
fatal: No names found, cannot describe anything.
/bin/bash: ../Scripts/osx_clt.sh: /bin/bash^M: bad interpreter: No such file or directory
mpiifx -c -m64 -fc=ifort -O2 -ipo -no-wrap-margin -fpp -DGITHASH_PP=\"-master\" -DGITDATE_PP=\""Sat Nov 9 09:28:05 2024 -0500\"" -DBUILDDATE_PP=\""Nov 11, 2024  16:31:37\""    -DUSE_IFPORT  -diag-disable=10448 ../../Source/prec.f90
mpiifx -c -m64 -fc=ifort -O2 -ipo -no-wrap-margin -fpp -DGITHASH_PP=\"-master\" -DGITDATE_PP=\""Sat Nov 9 09:28:05 2024 -0500\"" -DBUILDDATE_PP=\""Nov 11, 2024  16:31:37\""    -DUSE_IFPORT  -diag-disable=10448  ../../Source/imkl.f90
/bin/bash: mpiifx: command not found
make: *** [../makefile:151: prec.o] Error 127
make: *** Waiting for unfinished jobs....
/bin/bash: mpiifx: command not found
make: *** [../makefile:151: imkl.o] Error 127

STEP 3-3:
I use ‘vim ~/.bashrc’, and add
# oneAPI setup
source /opt/intel/oneapi/setvars.sh >& /dev/null

and also
export FIREMODELS=~/firemodels

at the last lines or the .bashrc file. Nonetheless, the error message in STEP 3-1 remains.
--------------------

I am struggling with how to proceed with the FDS compilation. Could you give me some suggestions so that I can further try? Thank you very much in advance.

Mingcian

**Comment by marcosvanella** [comment]
Hi, have you cloned the fds repo directly into your HPC cluster or copying it from somewhere else?

**Comment by marcosvanella** [comment]
Clone the repo for fds wherever you are going to compile it, load oneapi and try the Build/impi_intel_linux/make_fds.sh script.
Don't do this, the repo is already in unix EOL format:

_I use ‘dos2unix make_fds.sh’ to convert file make_fds.sh to Unix format._

**Reply by cxp484** [reply]
Also, to build release v6.9.1 instead of the latest source code, which includes additional development beyond v6.9.1, please check out that specific tag.
git checkout FDS-6.9.1

**Reply by mingcian3** [reply]
@marcosvanella:
Thanks for your reply. I add more details here:
1. I clone the fds, exp, out, smv directly from HPC, using (for example)
'git clone git@github.com:<username>/fds.git' 
which is suggested on https://github.com/firemodels/fds/wiki/Git-Notes-Getting-Started

2. If I don't use dos2unix make_fds.sh’ to convert file make_fds.sh to Unix format.
Then another error message shows:
'-bash: ./make_fds.sh: /bin/bash^M: bad interpreter: No such file or directory'

@cxp484:
Thanks for the reply. 
I clone the latest sourced on GitHub. 
I tried the command 'git checkout FDS-6.9.1' on HPC 
And an error message shows:
'error: pathspec 'FDS-6.9.1' did not match any file(s) known to git' 
Probably I don't use the command in the right place? 
Nonetheless, in my GitHub repositories, I have fetched the latest source, for fds, exp, out, and smv. For example, in 'Sync fork' the message shows:
'This branch is not behind the upstream firemodels/fds:master.
No new commits to fetch. Enjoy your day!'

Could you give me some advise that I could further test? Thanks!

**Reply by cxp484** [reply]
That is strange that you need to use dos2unix. I just checked the fds/Build/impi_intel_linux/make_fds.sh , and it seems to have unix line endings.  Don't clone it to a Windows machine first and then copy it to a linux machine. That might disturb the line endings. If possible clone directly in the linux machine. If you are doing so, I'm not sure from where the Windows line ending is coming from. 

Anyway, to checkout version 6.9.1 go to the fds folder and then checkout.
cd firemodels/fds
git checkout FDS-6.9.1

**Reply by mingcian3** [reply]
@cxp484:
Thanks for the suggestion. I go to ~/firemodels/fds and use the command
git checkout FDS-6.9.1
But the message shows:
'error: pathspec 'FDS-6.9.1' did not match any file(s) known to git'

Is there anything I could further try?

**Comment by cxp484** [comment]
Lets try this, I am not sure if your fork is up-to-date with firemodels/fds.

1. Update your fork with latest FDS repository:

cd firemodels/fds
git remote add firemodels git@github.com:firemodels/fds.git
git remote update
git merge firemodels/master

2. To build 

cd firemodels/fds/build/impi_intel_linux
./make_fds.sh

3. I am skipping the checkout 6.9.1 here, as the lastest code should build in your system. If you want to build release 6.9.1 then use following command before "2. Too build":

cd firemodels/fds
git checkout FDS-6.9.1

**Comment by cxp484** [comment]
Looking again to your log, it seems like the intel compiler setup in your cluster doesn't allow mpiifx -fc=ifort. Can you try following options in order:
1) cd firemodels/fds/Build/impi_intel_linux
    export INTEL_IFORT=ifx
    ./make_fds.sh

If above is unsuccessful, try following:
2) edit makefile to change compiler to mpiifort
    
    cd firemodels/fds/Build/impi_intel_linux
    sed -i '' "s/impi_intel_linux : FCOMPL = mpiifx/impi_intel_linux : FCOMPL = mpiifort/g" ../makefile
    ./make_fds.sh

**Reply by mingcian3** [reply]
@cxp484: 
Thank you very much for the suggestions. I tried all the commands on HPC, but the compilation remain failing. Here are the steps I've tried along with the error messages.

Test 1:
I try ‘ ./make_fds.sh’
Error message:
-bash: ./make_fds.sh: /bin/bash^M: bad interpreter: No such file or directory
Then I use ‘ dos2unix make_fds.sh’

Test 2:
I tried the commands:
cd firemodels/fds
git remote add firemodels git@github.com:firemodels/fds.git
git remote update
git merge firemodels/master
The final message: ‘Already up to date.’

Then I use the compile command ‘./make_fds.sh’.
The error message shows:

../Scripts/build_thirdparty_libs.sh: line 2: $'\r': command not found
../Scripts/build_thirdparty_libs.sh: line 6: syntax error near unexpected token `fi'
'./Scripts/build_thirdparty_libs.sh: line 6: `fi
Building impi_intel_linux
/bin/bash: ../Scripts/osx_clt.sh: /bin/bash^M: bad interpreter: No such file or directory
mpiifx -c -m64 -fc=ifx -O2 -ipo -no-wrap-margin -fpp -DGITHASH_PP=\"FDS-6.9.1-1275-g421fce2-master\" -DGITDATE_PP=\""Sat Nov 9 09:28:05 2024 -0500\"" -DBUILDDATE_PP=\""Nov 12, 2024  11:31:10\""    -DUSE_IFPORT  -diag-disable=10448 ../../Source/prec.f90
mpiifx -c -m64 -fc=ifx -O2 -ipo -no-wrap-margin -fpp -DGITHASH_PP=\"FDS-6.9.1-1275-g421fce2-master\" -DGITDATE_PP=\""Sat Nov 9 09:28:05 2024 -0500\"" -DBUILDDATE_PP=\""Nov 12, 2024  11:31:10\""    -DUSE_IFPORT  -diag-disable=10448  ../../Source/imkl.f90
Error: can not overwrite predefined compiler - ifx!
You should use mpiifort
ex. mpiifort -fc=ifx
make: *** [../makefile:151: prec.o] Error 1
make: *** Waiting for unfinished jobs....
Error: can not overwrite predefined compiler - ifx!
You should use mpiifort
ex. mpiifort -fc=ifx
make: *** [../makefile:151: imkl.o] Error 1


Test 3:
I tried ‘export INTEL_IFORT=ifx’
Then ‘./make_fds.sh’
Error message: 

../Scripts/build_thirdparty_libs.sh: line 2: $'\r': command not found
../Scripts/build_thirdparty_libs.sh: line 6: syntax error near unexpected token `fi'
'./Scripts/build_thirdparty_libs.sh: line 6: `fi
Building impi_intel_linux
/bin/bash: ../Scripts/osx_clt.sh: /bin/bash^M: bad interpreter: No such file or directory
mpiifx -c -m64 -fc=ifx -O2 -ipo -no-wrap-margin -fpp -DGITHASH_PP=\"FDS-6.9.1-1275-g421fce2-master\" -DGITDATE_PP=\""Sat Nov 9 09:28:05 2024 -0500\"" -DBUILDDATE_PP=\""Nov 12, 2024  11:46:45\""    -DUSE_IFPORT  -diag-disable=10448 ../../Source/prec.f90
mpiifx -c -m64 -fc=ifx -O2 -ipo -no-wrap-margin -fpp -DGITHASH_PP=\"FDS-6.9.1-1275-g421fce2-master\" -DGITDATE_PP=\""Sat Nov 9 09:28:05 2024 -0500\"" -DBUILDDATE_PP=\""Nov 12, 2024  11:46:45\""    -DUSE_IFPORT  -diag-disable=10448  ../../Source/imkl.f90
Error: can not overwrite predefined compiler - ifx!
You should use mpiifort
ex. mpiifort -fc=ifx
Error: can not overwrite predefined compiler - ifx!
You should use mpiifort
ex. mpiifort -fc=ifx
make: *** [../makefile:151: prec.o] Error 1
make: *** Waiting for unfinished jobs....
make: *** [../makefile:151: imkl.o] Error 1

Test 4:
I go to ‘firemodels/fds/Build/impi_intel_linux’
Try the command line: ‘sed -i '' "s/impi_intel_linux : FCOMPL = mpiifx/impi_intel_linux : FCOMPL = mpiifort/g" ../makefile’
The message:
sed: can't read s/impi_intel_linux : FCOMPL = mpiifx/impi_intel_linux : FCOMPL = mpiifort/g: No such file or directory

Could you give more advice that I could further try? Thanks a lot.

**Comment by rmcdermo** [comment]
Go to your fds repository and do a `git status -uno` and report the result.  There should be no reason for you to mess with line endings.  All the line endings in fds are for linux.

**Reply by mingcian3** [reply]
@rmcdermo: 
I go to ~/firemodels/fds, and use the command 'git status -uno'. 
The messages are:
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   Build/impi_intel_linux/make_fds.sh
        modified:   Build/impi_intel_linux_openmp/make_fds.sh
        modified:   Build/ompi_gnu_linux/make_fds.sh

no changes added to commit (use "git add" and/or "git commit -a")

**Reply by mingcian3** [reply]
Additional details of the HPC infrastructure: 

Recently the HPC infrastructure of our university is being updated from REHL 8 to REHL 9. However, the cluster I am using has not yet being updated and remains to be REHL 8. 
The cluster I am using is built with a 2x 48-core AMD EPYC 7552 processor. The operating system is still RHEL 8. I use the intel compiler version impi/2021.10.0-intel-compilers-2023.2.1.
I don’t know if the problem is caused by the AMD processor, where I am using the intel compiler, or the recent update of the operating system . So, I try to compile the previously cloned FDS6.8.0 on my HPC account. 

I go to ~/Build/impi_intel_linux.
I use the command ./make_fds.sh, and some error message shows:

../Scripts/set_intel_compiler.sh: line 3: $'\r': command not found
../Scripts/set_intel_compiler.sh: line 17: syntax error: unexpected end of file
Building impi_intel_linux with Intel MPI and
fatal: No names found, cannot describe anything.
/bin/bash: ../Scripts/intelmpi_compversion.sh: /bin/bash^M: bad interpreter: No such file or directory
/bin/bash: ../Scripts/openmpi_compversion.sh: /bin/bash^M: bad interpreter: No such file or directory
/bin/bash: ../Scripts/gnu_compversion.sh: /bin/bash^M: bad interpreter: No such file or directory
/bin/bash: ../Scripts/gnu_compmajorversion.sh: /bin/bash^M: bad interpreter: No such file or directory

Nonetheless, the compilation started afterwards and the compilation was successful. Therefore, I assume the problem is not caused by the AMD processor, the intel compiler version, or the operating system.

**Comment by rmcdermo** [comment]
I would go to your repo and get it back up to date.
```
cd fds
git checkout .
```
Then exit and relogin and compile the `opmi_gnu_linux` target.  Your hpc system should have instructions for loading modules for OpenMPI and GNU.

**Comment by mingcian3** [comment, marked as answer]
@rmcdermo , @marcosvanella , @cxp484 :

Thanks a lot for your help! Now I can compile and run simulation with the new executable. 

Here is how I did:
I load the modules for OpenMPI and GNU ('gompi/2023b' in our HPC).
I go to 'opmi_gnu_linux' and compile (./make_fds.sh). This time 'dos2unix make_fds.sh' is no longer needed and the compilation starts.

Thanks again for your help and suggestions. :)
